File: infrastructures/ddd/csharp/src/Infrastructure/Queues/Infra/IEventSubscriber.cs
using System;
using System.Threading.Tasks;

namespace OddlyDdd.Infrastructure.Queues.Infra
{
    /// <summary>
    /// Interface for subscribing to domain events from a message queue.
    /// Used for asynchronous communication between subdomains.
    /// Each subdomain subscribes to events from other subdomains it needs to react to.
    /// </summary>
    public interface IEventSubscriber
    {
        /// <summary>
        /// Subscribes to events on the specified topic with the provided handler.
        /// Topic naming convention: {subdomain}.{action}
        /// Examples: "example.created", "example.updated", "example.deleted"
        /// </summary>
        /// <typeparam name="TEvent">The event type</typeparam>
        /// <param name="p_topic">The topic to subscribe to</param>
        /// <param name="p_handler">The handler function to process events</param>
        Task SubscribeAsync<TEvent>(string p_topic, Func<TEvent, Task> p_handler);
    }
}

File: infrastructures/ddd/typescript/src/infrastructure/queues/infra/IEventSubscriber.ts
/**
 * Interface for subscribing to domain events from a message queue.
 * Event subscribers:
 * - Subscribe to topics to receive domain events
 * - Handle events from other subdomains
 * - Process events asynchronously
 * 
 * Subscribers live in /infrastructure/queues/subscribers/
 * Each subscriber handles events from a specific topic.
 */
export interface IEventSubscriber {
  /**
   * Subscribes to a topic and handles events with the provided handler.
   * @param p_topic The topic to subscribe to
   * @param p_handler The handler function to process events
   */
  SubscribeAsync<TEvent>(
    p_topic: string,
    p_handler: (p_event: TEvent) => Promise<void>
  ): Promise<void>;
}
